File: baseLib/src/androidTest/java/me/ycdev/android/lib/common/internalapi/android/os/PowerManagerIATest.java
package me.ycdev.android.lib.common.internalapi.android.os;

import android.content.Context;
import android.os.IBinder;

import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class PowerManagerIATest {
    @Test
    public void test_asInterface() {
        IBinder binder = ServiceManagerIA.getService(Context.POWER_SERVICE);
        assertNotNull(binder);

        Object service = PowerManagerIA.asInterface(binder);
        assertNotNull(service);
    }

    @Test
    public void test_getIPowerManager() {
        assertNotNull(PowerManagerIA.getIPowerManager());
    }

    @Test
    public void test_reboot() {
        assertTrue(PowerManagerIA.checkReflect_reboot());
    }

    @Test
    public void test_shutdown() {
        assertTrue(PowerManagerIA.checkReflect_shutdown());
    }

    @Test
    public void test_crash() {
        assertTrue(PowerManagerIA.checkReflect_crash());
    }

    @Test
    public void test_goToSleep() {
        assertTrue(PowerManagerIA.checkReflect_goToSleep());
    }
}

File: baseLib/src/androidTest/java/me/ycdev/android/lib/common/internalapi/android/os/PowerManagerIATest.kt
package me.ycdev.android.lib.common.internalapi.android.os

import android.content.Context
import android.os.Build
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Assume.assumeTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class PowerManagerIATest {
    @Test
    fun test_asInterface() {
        val binder = ServiceManagerIA.getService(Context.POWER_SERVICE)
        assertNotNull(binder)

        val service = PowerManagerIA.asInterface(binder!!)
        assertNotNull(service)
    }

    @Test
    fun test_getIPowerManager() {
        assertNotNull(PowerManagerIA.iPowerManager)
    }

    @Test
    fun test_reboot() {
        assertTrue(PowerManagerIA.checkReflectReboot())
    }

    @Test
    fun test_shutdown() {
        assumePowerManagerHiddenApiReflectionAllowed()
        assertTrue("failed to reflect #shutdown", PowerManagerIA.checkReflectShutdown())
    }

    @Test
    fun test_crash() {
        assumePowerManagerHiddenApiReflectionAllowed()
        assertTrue("failed to reflect #crash", PowerManagerIA.checkReflectCrash())
    }

    @Test
    fun test_goToSleep() {
        assumePowerManagerHiddenApiReflectionAllowed()
        assertTrue("failed to reflect #goToSleep", PowerManagerIA.checkReflectGoToSleep())
    }

    private fun assumePowerManagerHiddenApiReflectionAllowed() {
        assumeTrue(
            "PowerManager hidden API reflection is blocked on Android P and newer",
            Build.VERSION.SDK_INT < Build.VERSION_CODES.P
        )
    }
}
